src/Compiler/Microsoft.AspNetCore.Razor.Language/src/TagHelperDescriptorBuilderExtensions.cs

```diff
@@ -4,11 +4,92 @@
 #nullable disable
 
 using System;
+using System.Collections.Generic;
+using System.Diagnostics;
 
 namespace Microsoft.AspNetCore.Razor.Language;
 
 public static class TagHelperDescriptorBuilderExtensions
 {
+    public static void SetMetadata(
+        this TagHelperDescriptorBuilder builder,
+        KeyValuePair<string, string> pair)
+    {
+        if (pair.Key == TagHelperMetadata.Runtime.Name)
+        {
+            builder.SetMetadata(MetadataCollection.Create(pair));
+        }
+        else
+        {
+            builder.SetMetadata(
+                MetadataCollection.Create(
+                    pair,
+                    new(TagHelperMetadata.Runtime.Name, TagHelperConventions.DefaultKind)));
```

Comment:
`MetadataHolder` will throw if it has to add the runtime name metadata to an existing `MetadataCollection` to avoid allocating a new `MetadataCollection`. (There's a comment in `MetadataHolder.AddIfMissing` that describes this behavior.) So, I made sure that these helpers always add the value for tag helper descriptors.